### PR TITLE
Unifying the recommenders #58

### PR DESCRIPTION
**Pull request by mostafa-mahmoud**

### CODE REVIEW

**Comment by mostafa-mahmoud** [comment]
@ndomar I moved rounded_predictions, recommend_items and get_ratings to the AbstractRecommender.

In the recommend_items, I used get_predictions() instead of predictions:
user_ratings = self.get_predictions()[user_id]

The rest shouldn't interfere with the logic that much.

**Comment by mostafa-mahmoud** [comment]
@ndomar  Also, I see that you don't cache the predictions matrix #57

**Comment by mostafa-mahmoud** [comment]
There's a failing test in evaluator_tests line 60 (recall test)

File: util/recommender_configuer.py [hunk 1]
```diff
@@ -39,6 +39,15 @@ def get_collaborative_filtering(self):
         """
         return self.config_dict['collaborative-filtering']
 
+    def get_recommender(self):
+        """
+        Get the recommender type
+
+        :returns: A string of userbased or itembased recommendations.
+        :rtype: str
```
Comment (ndomar) [inline util/recommender_configuer.py]: It returns a dict?

File: lib/content_based.py [hunk 2]
```diff
@@ -63,49 +73,26 @@ def get_document_topic_distribution(self):
...
         # predicted_rating[u,i] = sum[j]{R[u,j] Vj * Vi} / sum[j]{Vj * Vi}
         #                       = sum[j]{R[u,j] * cos(i, j)} / sum[j]{cos(i, j)}
-        if self.predictions is not None:
-            return self.predictions
-
         V = self.document_distribution.copy()
         for item in range(V.shape[0]):
-            V[item] /= numpy.sqrt(V[item].dot(V[item]))
-        self.predictions = self.ratings.dot(V).dot(V.T) / V.dot(V.T.dot(numpy.ones((V.shape[0],))))
+            mag = numpy.sqrt(V[item].dot(V[item]))
```
Comment (ndomar) [inline lib/content_based.py]: better variable name than mag

File: util/recommender_configuer.py (hunk 1 above)
Comment (mostafa-mahmoud) [inline util/recommender_configuer.py]: No, this is a string of 'itembased' or 'usedbased', based upon, we will use the results of contentbased or collaborative-filtering for the predictions.

**Pull request merged**